### PR TITLE
Command termination on timeout

### PR DESCRIPTION
**Pull request by marcin-usielski**

### CODE REVIEW

File: moler/connection_observer.py [hunk 1]
```diff
@@ -243,13 +246,28 @@ def data_received(self, data):
...
+        """
+        Should be used to indicate some failure during observation.
+
+        :param exception: Exception to set
+        :return: None
+        """
+        self._set_exception(exception)
+        self._is_done = True
+
+    def _set_exception(self, exception):
```
Comment (Ernold11) [inline moler/connection_observer.py]: maybe _set_exception_without_done?

Comment (marcin-usielski) [reply moler/connection_observer.py]: Chnaged.

File: moler/connection_observer.py [hunk 2]
```diff
@@ -243,13 +246,28 @@ def data_received(self, data):
...
+        Should be used to indicate some failure during observation. This method does not finish connection observer
+        object!
+
+        :param exception: exception to set
+        :return: None
+        """
         if self._is_done:
             self._log(logging.WARNING,
-                      "Trial to set exception {!r} on already done {}".format(exception, self),
+                      "Try to set exception {!r} on already done {}".format(exception, self),
```
Comment (marcin-usielski) [inline moler/connection_observer.py]: Check the spelling.

File: moler/runner.py [hunk 3]
```diff
@@ -345,13 +346,13 @@ def wait_for(self, connection_observer, connection_observer_future, timeout=None
                     if (future in done) or connection_observer.done():
                         self._cancel_submitted_future(connection_observer, future)
                         return None
-                    timeout = connection_observer.timeout
+                    timeout = connection_observer.timeout + connection_observer.terminating_timeout
```
Comment (marcin-usielski) [inline moler/runner.py]: eol_timeout = = connection_observer.timeout + connection_observer.terminating_timeout

File: moler/runner.py [hunk 4]
```diff
@@ -345,13 +346,13 @@ def wait_for(self, connection_observer, connection_observer_future, timeout=None
...
                     already_passed = time.time() - start_time
                     remain_time = timeout - already_passed
```
Comment (marcin-usielski) [inline moler/runner.py]: remain_time = eol_timeout - already_passed.

Comment (marcin-usielski) [reply moler/runner.py]: Inside loop call  _wait_for_time_out (remove future cancel and connection_observer.cncel) and after loop call _remove_future

File: moler/runner.py [hunk 5]
```diff
@@ -374,6 +375,7 @@ def _wait_for_time_out(self, connection_observer, connection_observer_future, ti
             time_out_observer(connection_observer=connection_observer,
                               timeout=timeout, passed_time=passed,
                               runner_logger=self.logger, kind="await_done")
+        connection_observer.cancel()
```
Comment (marcin-usielski) [inline moler/runner.py]: to remove

File: moler/runner.py [hunk 6]
```diff
@@ -466,13 +469,27 @@ def _feed_loop(self, connection_observer, stop_feeding, observer_lock):
...
-                                      timeout=connection_observer.timeout,
-                                      passed_time=run_duration,
-                                      runner_logger=self.logger)
-                break
+            timeout = connection_observer.timeout
+            if in_terminating:
+                timeout = connection_observer.terminating_timeout
+            if (timeout is not None) and (run_duration >= timeout):
+                if in_terminating:
+                    msg = "{} could not terminate during {} seconds. It will be cancelled".format(connection_observer,
```
Comment (marcin-usielski) [inline moler/runner.py]: It will be forcefully terminated.

Comment (marcin-usielski) [reply moler/runner.py]: Underlying real command could not be finished.

**Pull request merged**